## Sam Weekend Example


In this example we are introduced to Sam. Sam is either healthy or sick, this is thus our state space. Sam can choose to either party or rest over the weekend, thus defining our action space. Sam estimates the probabilities and rewards related to their state and action. We can model the following problem as a MDP as shown below: 

### Model

State Set = { Healthy, Sick }

Action Set = { Relax, Party }

Probabilitiy + Reward Functions:

| s  | a | s’ | P(s’\|s,a) | R(s,a) |
|:--:|:--:|:--:|:----------:|:----------:|
| Healthy | Relax | Healthy| 0.95        | 7         |
| Healthy | Relax | Sick   | 0.05       | 7         |
| Healthy | Party | Healthy| 0.7        | 10         |
| Healthy | Party | Sick   | 0.3        | 10         |
| Sick    | Relax | Healthy| 0.5        | 0       |
| Sick    | Relax | Sick   | 0.5        | 0         |
| Sick    | Party | Healthy| 0.1        | 2         |
| Sick    | Party | Sick   | 0.9        | 2        |

To be eligible for our function, we need to present this information as two lists and two functions, which is shown below:

In [1]:
S1=["healthy","sick"]
A1=["relax","party"]
def R1(sdash,s,a):
    if s == "healthy":
        if a == "relax":
            return 7
        else:
            return 10
    else:
        if a == "relax":
            return 0
        else:
            return 2
def P1(sdash,s,a):
    if s == "healthy":
        if a == "relax":
            p=0.95
        else:
            p=0.7
    else:
        if a == "relax":
            p=0.5
        else:
            p=0.1
    if sdash == "healthy":
        return p
    else:
        return 1-p

Since our reward function does not takes s', we want to use mode 1 of our function. We then pick the usual $\gamma=0.9$ and set $\epsilon=0.00001$, as we want to have a really low epsilon to make sure we have converged to the correct values. We can now import our Value_iteration function from our package and run it.

In [4]:
from ValueIt.ValueItFunction import Value_iteration 

Value_iteration(S1,A1,P1,R1,0.9,0.00001,1)

[['party', 'relax'], [67.07308500982839, 54.87796305860887]]

The optimal policy for each state and associated value can be found in the table below:

| State | Optimal Policy | Value |
|:-----:|:--------------:|:-----:|
|Healthy| Party          | 67.073|
| Sick  | Relax          | 54.878|